### PR TITLE
UniversalRP increase light limits

### PR DESCRIPTION
**Pull request by phi-lira**
### Purpose of this PR
 Increases light limits for UniversalRP. 

 New limits: 
    Per-object lights: 8, except in GLES2 that is 4.
    Visible Lights: 256 if not mobile, 32 otherwise.

 Non-mobile path store lights in StructuredBuffer.
 Currently there are some performance issues with StructuredBuffer in mobile. It uses UBO path instead.

 Performance Test on Android. No regression found
 There's some improvement in performance in some Android GLES3 devices.

---
### Release Notes
 Added
---
### Testing status
**Katana Tests**: First off we need to make sure the Katana SRP tests are green?

**Manual Tests**: 
  - Tested UniversalGraphicsTest project in GLES3, Vulkan, Metal and Nintendo Switch. No regressions found.
  - Performance Testing on GLES3 Potato Benchmark. Same or better performance.
  - QA validated template and BoatAttack

- [x ] Opened test project + Run graphic tests locally
- [x ] Built a player
- [ ] Checked new UI names with UX convention
- [ ] Tested UI multi-edition + Undo/Redo
- [x ] C# and shader warnings (supress shader cache to see them)
- [ ] Checked new resources path for the reloader (in devloper mode, you have a button at end of resources that check the pathes)
- Other: 

**Automated Tests**: 
 - Added two new automated tests. 
    008 - Tests up to 8 per-object lights
    009 - Tests directional lights + 32 additional lights casting shadows.

Launch Katana (Link on master here - update for your branch):
https://katana.bf.unity3d.com/projects/com.unity.render-pipelines/builders?ScriptableRenderLoop_branch=lw/increase-light-limits&automation-tools_branch=master&unity_branch=trunk&sort=1-asc

Alternative, launch Yamato (Select your branch):
https://yamato.prd.cds.internal.unity3d.com/jobs/78-ScriptableRenderPipeline

Any test projects to go with this to help reviewers?

---
### Overall Product Risks
**Technical Risk**: High. Changes how lighting data is setup in UniversalRP.

**Halo Effect**: High. Touches all platforms in different ways.

---
### Comments to reviewers
Notes for the reviewers you have assigned.

### CODE REVIEW

File: com.unity.render-pipelines.universal/Runtime/ForwardLights.cs [hunk 1]
```diff
@@ -35,23 +38,37 @@ static class LightConstantBuffer
         Vector4[] m_AdditionalLightSpotDirections;
         Vector4[] m_AdditionalLightOcclusionProbeChannels;
 
+        bool m_UseStructuredBuffer;
+        
         public ForwardLights()
         {
+            GraphicsDeviceType gfxDeviceType = SystemInfo.graphicsDeviceType;
```
Comment (aleks01010101) [inline com.unity.render-pipelines.universal/Runtime/ForwardLights.cs]: Unused?

File: com.unity.render-pipelines.universal/Runtime/ForwardLights.cs [hunk 2]
```diff
@@ -180,22 +197,14 @@ void InitializeLightConstants(NativeArray<VisibleLight> lights, int lightIndex,
...
-                    out m_AdditionalLightColors[i],
-                    out m_AdditionalLightAttenuations[i],
-                    out m_AdditionalLightSpotDirections[i],
-                    out m_AdditionalLightOcclusionProbeChannels[i]);
-
             m_MixedLightingSetup = MixedLightingSetup.None;
 
             // Main light has an optimized shader path for main light. This will benefit games that only care about a single light.
-            // Universal pipeline also supports only a single shadow light, if available it will be the main light.
+            // Lightweight pipeline also supports only a single shadow light, if available it will be the main light.
```
Comment (aleks01010101) [inline com.unity.render-pipelines.universal/Runtime/ForwardLights.cs]: That's probably an unintentional change... :)

File: com.unity.render-pipelines.universal/Runtime/ForwardLights.cs [hunk 3]
```diff
@@ -207,25 +216,62 @@ void SetupMainLightConstants(CommandBuffer cmd, ref LightData lightData)
...
+                    for (int i = 0, lightIter = 0; i < lights.Length && lightIter < maxAdditionalLightsCount; ++i)
                     {
-                        InitializeLightConstants(lights, i, out m_AdditionalLightPositions[additionalLightsCount],
-                            out m_AdditionalLightColors[additionalLightsCount],
-                            out m_AdditionalLightAttenuations[additionalLightsCount],
-                            out m_AdditionalLightSpotDirections[additionalLightsCount],
-                            out m_AdditionalLightOcclusionProbeChannels[additionalLightsCount]);
-                        additionalLightsCount++;
+                        VisibleLight light = lights[i];
+                        if (lightData.mainLightIndex != i && light.lightType != LightType.Directional)
```
Comment (aleks01010101) [inline com.unity.render-pipelines.universal/Runtime/ForwardLights.cs]: Isn't mainLight always directional?

File: com.unity.render-pipelines.universal/Runtime/Passes/AdditionalLightsShadowCasterPass.cs [hunk 4]
```diff
@@ -170,59 +215,131 @@ void RenderAdditionalShadowmapAtlas(ref ScriptableRenderContext context, ref Cul
...
+
+                cmd.SetGlobalMatrixArray(AdditionalShadowsConstantBuffer._AdditionalLightsWorldToShadow, additionalLightShadowMatrices.ToArray());
+                cmd.SetGlobalFloatArray(AdditionalShadowsConstantBuffer._AdditionalShadowStrength, m_AdditionalLightsShadowStrength);
+
+                additionalLightShadowMatrices.Dispose();
+            }
+
+            if (softShadows)
+            {
+                // Currently only used when SHADER_API_MOBILE
```
Comment (aleks01010101) [inline com.unity.render-pipelines.universal/Runtime/Passes/AdditionalLightsShadowCasterPass.cs]: This looks like we can skip this based on mobile/not

File: com.unity.render-pipelines.universal/Runtime/Passes/MainLightShadowCasterPass.cs [hunk 5]
```diff
@@ -203,20 +205,41 @@ void SetupMainLightShadowReceiverConstants(CommandBuffer cmd, ref ShadowData sha
...
+                cmd.SetGlobalVector(MainLightShadowConstantBuffer._CascadeShadowSplitSphereRadii, new Vector4(
+                    m_CascadeSplitDistances[0].w * m_CascadeSplitDistances[0].w,
+                    m_CascadeSplitDistances[1].w * m_CascadeSplitDistances[1].w,
+                    m_CascadeSplitDistances[2].w * m_CascadeSplitDistances[2].w,
+                    m_CascadeSplitDistances[3].w * m_CascadeSplitDistances[3].w));
+            }
+
+            if (softShadows)
+            {
+                // Currently only used when SHADER_API_MOBILE
```
Comment (aleks01010101) [inline com.unity.render-pipelines.universal/Runtime/Passes/MainLightShadowCasterPass.cs]: ...and here as well

File: com.unity.render-pipelines.universal/Runtime/UniversalRenderPipeline.cs [hunk 6]
```diff
@@ -65,14 +65,20 @@ public static float maxRenderScale
         // This uses unity_4LightIndices to store an array of 4 light indices
```
Comment (aleks01010101) [inline com.unity.render-pipelines.universal/Runtime/UniversalRenderPipeline.cs]: I guess this comment is outdated

File: com.unity.render-pipelines.universal/ShaderLibrary/Input.hlsl [hunk 7]
```diff
@@ -1,19 +1,16 @@
 #ifndef UNIVERSAL_INPUT_INCLUDED
 #define UNIVERSAL_INPUT_INCLUDED
 
-#define MAX_VISIBLE_LIGHTS 16
+#define MAX_VISIBLE_LIGHTS 64
```
Comment (aleks01010101) [inline com.unity.render-pipelines.universal/ShaderLibrary/Input.hlsl]: Shouldn't it be 1024 for non-mobile?

File: com.unity.render-pipelines.universal/ShaderLibrary/Lighting.hlsl [hunk 8]
```diff
@@ -170,6 +165,38 @@ Light GetAdditionalLight(int i, float3 positionWS)
...
+    // The cast below will be done inside the light loop if this function is called from GetAdditionalLight below.
+    // This is bad for performance.
+    int perObjectIndices[MAX_PEROBJECT_LIGHTS] = (int[MAX_PEROBJECT_LIGHTS])unity_LightIndices;
+    return perObjectIndices[index];
+#endif
+}
+
+// Fills a light struct given a loop i index. This will convert the i
+// index to a perObjectLightIndex
+// Note: BEGIN_LIGHT_LOOP / END_LIGHT_LOOP instead
```
Comment (aleks01010101) [inline com.unity.render-pipelines.universal/ShaderLibrary/Lighting.hlsl]: Note: Use ...

File: com.unity.render-pipelines.universal/ShaderLibrary/Input.hlsl [hunk 9]
```diff
@@ -30,27 +27,36 @@ struct InputData
...
 half4 _AdditionalLightsCount;
+#if !USE_STRUCTURED_BUFFER_FOR_LIGHT_DATA
 float4 _AdditionalLightsPosition[MAX_VISIBLE_LIGHTS];
 half4 _AdditionalLightsColor[MAX_VISIBLE_LIGHTS];
 half4 _AdditionalLightsAttenuation[MAX_VISIBLE_LIGHTS];
 half4 _AdditionalLightsSpotDir[MAX_VISIBLE_LIGHTS];
 half4 _AdditionalLightsOcclusionProbes[MAX_VISIBLE_LIGHTS];
-CBUFFER_END
+#endif
+
```
Comment (aleks01010101) [inline com.unity.render-pipelines.universal/ShaderLibrary/Input.hlsl]: Why not just use `#else` here?

File: com.unity.render-pipelines.universal/Runtime/ForwardLights.cs (hunk 3 above)
Comment (phi-lira) [reply com.unity.render-pipelines.universal/Runtime/ForwardLights.cs]: Yes, but this is also removing additional directional lights.

File: com.unity.render-pipelines.universal/ShaderLibrary/Input.hlsl (hunk 7 above)
Comment (phi-lira) [reply com.unity.render-pipelines.universal/ShaderLibrary/Input.hlsl]: non-mobile doesn't use this define. It uses StructuredBuffer.

Comment (aleks01010101) [reply com.unity.render-pipelines.universal/ShaderLibrary/Input.hlsl]: Then it has to agree with the comment in C# (which says "Keep in sync with hlsl") :)

File: com.unity.render-pipelines.universal/ShaderLibrary/Input.hlsl (hunk 9 above)
Comment (phi-lira) [reply com.unity.render-pipelines.universal/ShaderLibrary/Input.hlsl]: Fixed

File: com.unity.render-pipelines.universal/ShaderLibrary/Lighting.hlsl (hunk 8 above)
Comment (phi-lira) [reply com.unity.render-pipelines.universal/ShaderLibrary/Lighting.hlsl]: Fixed

File: com.unity.render-pipelines.universal/Runtime/UniversalRenderPipeline.cs (hunk 6 above)
Comment (phi-lira) [reply com.unity.render-pipelines.universal/Runtime/UniversalRenderPipeline.cs]: Fixed

File: com.unity.render-pipelines.universal/Runtime/Passes/MainLightShadowCasterPass.cs (hunk 5 above)
Comment (phi-lira) [reply com.unity.render-pipelines.universal/Runtime/Passes/MainLightShadowCasterPass.cs]: Fixed.

File: com.unity.render-pipelines.universal/Runtime/Passes/AdditionalLightsShadowCasterPass.cs (hunk 4 above)
Comment (phi-lira) [reply com.unity.render-pipelines.universal/Runtime/Passes/AdditionalLightsShadowCasterPass.cs]: Fixed.

File: com.unity.render-pipelines.universal/Runtime/ForwardLights.cs (hunk 2 above)
Comment (phi-lira) [reply com.unity.render-pipelines.universal/Runtime/ForwardLights.cs]: Fixed.

File: com.unity.render-pipelines.universal/Runtime/ForwardLights.cs (hunk 1 above)
Comment (phi-lira) [reply com.unity.render-pipelines.universal/Runtime/ForwardLights.cs]: Fixed.

File: com.unity.render-pipelines.universal/Runtime/ForwardLights.cs (hunk 3 above)
Comment (aleks01010101) [reply com.unity.render-pipelines.universal/Runtime/ForwardLights.cs]: You could just leave `light.lightType != LightType.Directional` then

**Comment by Katstodian** [comment]
Updated the Comparison page with new light limits. @phi-lira, could you check and confirm?

File: com.unity.render-pipelines.universal/Runtime/ShaderData.cs [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+        internal ComputeBuffer GetShadowIndicesBuffer(int size)
+        {
+            return GetOrUpdateBuffer<int>(ref m_ShadowIndicesBuffer, size);
+        }
+
+        ComputeBuffer GetOrUpdateBuffer<T>(ref ComputeBuffer buffer, int size) where T : struct
+        {
+            if (buffer == null)
+            {
+                buffer = new ComputeBuffer(size, Marshal.SizeOf<T>());
```
Comment (pbbastian) [inline com.unity.render-pipelines.universal/Runtime/ShaderData.cs]: Use `UnsafeUtility.SizeOf` instead

File: com.unity.render-pipelines.universal/Runtime/ShaderData.cs [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+        ComputeBuffer m_LightIndicesBuffer = null;
+
+        ComputeBuffer m_ShadowDataBuffer = null;
+        ComputeBuffer m_ShadowIndicesBuffer = null;
+
+        ShaderData()
+        {
+        }
+
+        internal static ShaderData instance
```
Comment (pbbastian) [inline com.unity.render-pipelines.universal/Runtime/ShaderData.cs]: Hmm, I think this class would be better off as a normal class -- i.e. not a singleton. It seems like its life-time is very tied to the render pipeline instance anyway.

**Pull request merged**